### PR TITLE
Adds subgraph volume-computation, support for directed graphs

### PR DESCRIPTION
**Pull request by fabratu**
This PR would add the following:

- Adds support for volume of directed graphs. Implementation following initial thoughts from https://github.com/networkit/networkit/pull/558#issuecomment-632704544
- Adds support for volume of subgraphs. Implementation following initial thoughts from https://github.com/networkit/networkit/pull/558#discussion_r430247248
- Adds unit-tests and python-interface https://github.com/networkit/networkit/pull/558#pullrequestreview-421021520

### CODE REVIEW

**Comment by fabratu** [comment]
PR is ready for review (error in doc-build fixed in #590 )

File: networkit/cpp/graph/GraphTools.cpp [hunk 1]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
-    if (!G.isWeighted()) {
-        return 2.0 * G.numberOfEdges();
+double outVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double outVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() > 0 && G.numberOfNodes() != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            outVolume += G.weightedDegree(*it, directed);
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:241]: Counting self-loops twice for directed graphs seems wrong. The version for the whole graph counts them twice only for undirected graphs (as I would expect). I think this should be `!directed`. Or maybe not, as the variable is actually `true` if the graph is undirected?!

File: networkit/cpp/graph/GraphTools.cpp [hunk 2]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
-    if (G.isDirected())
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+double inVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double inVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() && G.numberOfNodes() != nodes.size()) {
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            inVolume += G.weightedDegreeIn(*it, directed);
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:210]: Counting self-loops twice for directed graphs seems wrong. The version for the whole graph counts them twice only for undirected graphs (as I would expect). I think this should be `!directed`. Or maybe not, as the variable is actually `true` if the graph is undirected?!

File: networkit/cpp/graph/GraphTools.cpp [hunk 3]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
+
+            // This removes all incoming edges from the rest of the graph to the subgraph
+            G.forInNeighborsOf(*it, [&](node v) {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    inVolume -= static_cast<double>(G.weight(v, *it));
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:216]: The edge weight should instead be taken as parameter in this lambda by defining it as `[&](node, node v, edgeweight w)`.

File: networkit/cpp/graph/GraphTools.cpp [hunk 4]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
 
-double volume(const Graph &G) {
-    if (G.isDirected())
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+double inVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double inVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() && G.numberOfNodes() != nodes.size()) {
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:208]: While the optimization for not counting the whole graph seems neat, I wonder if we should not check that the graph really contains all nodes that are in the set.

File: networkit/cpp/graph/GraphTools.cpp [hunk 5]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:209]: I would suggest a quick check if the node actually exists here as the performance impact should be minimal (given that we iterate over all neighbors, anyway).

File: networkit/cpp/graph/GraphTools.cpp [hunk 6]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
+            inVolume += G.weightedDegreeIn(*it, directed);
+
+            // This removes all incoming edges from the rest of the graph to the subgraph
+            G.forInNeighborsOf(*it, [&](node v) {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    inVolume -= static_cast<double>(G.weight(v, *it));
+                }
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:217]: For weighted graphs, this iterates over the neighbors twice: Once to collect the edge weights in `weightedDegree` and once here to remove some of them again. Why don't we directly only add those edge weights to neighbors that are in the set, i.e., remove the call to `G.weightedDegree` and add weights in the second loop when `neighborIt != nodes.end()`.

File: networkit/cpp/graph/GraphTools.cpp [hunk 7]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
+}
 
-    if (!G.isWeighted()) {
-        return 2.0 * G.numberOfEdges();
+double outVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double outVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
+
+    // Compute volume of subgraph
+    if (!nodes.empty() > 0 && G.numberOfNodes() != nodes.size()) {
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:239]: Same here: While the optimization for not counting the whole graph seems neat, I wonder if we should not check that the graph really contains all nodes that are in the set.

File: networkit/cpp/graph/GraphTools.cpp [hunk 8]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
+        for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
+            outVolume += G.weightedDegree(*it, directed);
+
+            // This removes all outgoing edges from the subgraph to the rest
+            G.forNeighborsOf(*it, [&](node v) {
+                std::unordered_set<node>::const_iterator neighborIt = nodes.find(v);
+                if (neighborIt == nodes.end()) {
+                    outVolume -= static_cast<double>(G.weight(*it, v));
+                }
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:248]: Same here: For weighted graphs, this iterates over the neighbors twice: Once to collect the edge weights in `weightedDegree` and once here to remove some of them again. Why don't we directly only add those edge weights to neighbors that are in the set, i.e., remove the call to `G.weightedDegree` and add weights in the second loop when `neighborIt != nodes.end()`.

File: networkit/cpp/graph/GraphTools.cpp [hunk 9]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
     return m / (n * (n - 1));
 }
 
-double volume(const Graph &G) {
-    if (G.isDirected())
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+double inVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double inVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:205]: Do you mean `bool directed = !G.isDirected()`? If yes, this is extremely confusing. Please rename the variable to `undirected`.

File: networkit/cpp/graph/GraphTools.cpp [hunk 10]
```diff
@@ -200,15 +200,66 @@ double density(const Graph &G) noexcept {
...
+        }
+    }
+    return inVolume;
+}
 
-    if (!G.isWeighted()) {
-        return 2.0 * G.numberOfEdges();
+double outVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double outVolume = 0.0;
+    bool directed = G.isDirected() ? false : true;
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:236]: Do you mean `bool directed = !G.isDirected()`? If yes, this is extremely confusing. Please rename the variable to `undirected`.

**Review by michitux — CHANGES_REQUESTED** [review]
Why is the method `volume` removed without any deprecation? Further, the term _volume_ of a subset of nodes as I know it is usually the sum of the weights of the edges incident to the nodes, see, e.g., [Wikipedia on Conductance](https://en.wikipedia.org/wiki/Conductance_(graph)).

This is two times almost the same code. I would suggest to refactor this code to just one implementation.

**Comment by fabratu** [comment]
Thanks for the thorough review, this helps a lot.

While most of it is clear, I would suggest leave two implementations. This split into out/in-functions is used in several places in nkt. However I changed the naming to align with other functions like `weightedDegree`. Also re-introduced the `volume`-function and split subgraph/graph-volume computation. Does it look better now?

File: networkit/cpp/graph/GraphTools.cpp [hunk 11]
```diff
@@ -201,14 +201,65 @@ double density(const Graph &G) noexcept {
...
+    return volume;
+}
+
+double subgraphInVolume(const Graph &G, const std::unordered_set<node> &nodes) {
+    double inVolume = 0.0;
+    bool countSelfLoopsTwice = !G.isDirected();
+    count nodesNotInGraph = 0;
+
+    // Check for nodes in subset, which are not part of the graph
+    for (auto it = std::begin(nodes); it != std::end(nodes); ++it) {
```
Comment (angriman) [inline networkit/cpp/graph/GraphTools.cpp:223]: What about `for (node u : nodes)`?

File: include/networkit/graph/GraphTools.hpp [hunk 12]
```diff
@@ -129,13 +129,36 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+/**
+ * Returns the volume (sum of the in-degree of all nodes) of a subgraph 
+ * defined by a set of nodes. If the set is empty, the volume of the complete
+ * graph is computed.
+ *
+ * @param G The input graph.
+ * @param nodes Subset of nodes.
+ * @return inVolume of the subgraph.
+ */
+double subgraphInVolume(const Graph &G, const std::unordered_set<node> &nodes);
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:149]: For the input nodes, can we use ranges? So the user is not forced to create an `unordered_set`.

File: include/networkit/graph/GraphTools.hpp [hunk 13]
```diff
@@ -129,13 +129,36 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+/**
+ * Returns the volume (sum of the out-degree of all nodes) of a subgraph 
+ * defined by a set of nodes. If the set is empty, the volume of the complete
+ * graph is computed.
+ *
+ * @param G The input graph.
+ * @param nodes Subset of nodes.
+ * @return Volume of the graph.
+ */
+double subgraphVolume(const Graph &G, const std::unordered_set<node> &nodes);
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:160]: Same here.

File: networkit/cpp/graph/GraphTools.cpp (hunk 11 above)
Comment (angriman) [reply networkit/cpp/graph/GraphTools.cpp:223]: Of course my comment does not hold anymore if we are going to use ranges for the input nodes.

File: networkit/cpp/graph/GraphTools.cpp [hunk 14]
```diff
@@ -201,14 +201,65 @@ double density(const Graph &G) noexcept {
...
+        if (!G.hasNode(*it)) {
+            nodesNotInGraph++;
+        } 
+    } 
+
+    // Compute volume of subgraph if node-subset is non-empty
+    if (!nodes.empty() > 0 && (G.numberOfNodes() - nodesNotInGraph) != nodes.size()) {
```
Comment (michitux) [inline networkit/cpp/graph/GraphTools.cpp:254]: The `> 0` in the first condition should be removed. The `nodesNotInGraph` should be subtracted from `nodes.size()`, not `G.numberOfNodes()`.

File: networkit/cpp/graph/test/GraphToolsGTest.cpp [hunk 15]
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
...
+        const auto G2 = GraphTools::subgraphFromNodes(G, sNodes);
+
+        const auto outVolumeSub = GraphTools::subgraphVolume(G, sNodes);
+        const auto inVolumeSub = GraphTools::subgraphInVolume(G, sNodes);
+
+        // Volume for either directed/undirected
+        double baseG2 = weighted() ? static_cast<double>(G2.totalEdgeWeight())
+                                   : static_cast<double>(G2.numberOfEdges());
+        EXPECT_NEAR(outVolumeSub, mod * baseG2, 1e-7);
+        EXPECT_NEAR(inVolumeSub, mod * baseG2, 1e-7);
```
Comment (michitux) [inline networkit/cpp/graph/test/GraphToolsGTest.cpp:132]: This does not match the new definition of volume and the test fails accordingly.

**Review by michitux — CHANGES_REQUESTED** [review]
This looks much better. I agree with @angriman concerning the use of ranges, in particular as with the new implementation, no set membership needs to be tested anymore. I am not so sure about the names of the functions, though, as this is not the volume of a subgraph, but the volume of a set of nodes. Maybe just leaving them as `volume`/`inVolume` makes more sense?

File: include/networkit/graph/GraphTools.hpp [hunk 16]
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
 double density(const Graph &G) noexcept;
 
 /**
- * Returns the volume (sum of the out-degree of all nodes) of the graph.
+ * Returns the volume of the input graph.
  *
  * @param G The input graph.
  * @return Volume of the graph.
+ *
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:136]: This change is not needed

File: include/networkit/graph/GraphTools.hpp [hunk 17]
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+ * 
  */
 double volume(const Graph &G);
 
+/**
+ * Returns the volume (sum of the out-degree of all nodes) of a range
+ * of nodes.
+ *
+ * @param G The input graph.
+ * @param first Iterator for the starting node in the set.
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:145]: Maybe it is clearer `@param first,last The range of nodes to insert` as they do in cppreference (e.g., https://en.cppreference.com/w/cpp/container/vector/insert).

File: include/networkit/graph/GraphTools.hpp [hunk 18]
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+ * @param last Iterator for the last node in the set.
+ * @return volume of the graph.
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:147]:
```suggestion
 * @return volume of the graph.
```
```suggestion
 * @return Volume of the graph.
```

File: include/networkit/graph/GraphTools.hpp [hunk 19]
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+ */
+template <class InputIt>
+double volume(const Graph &G, InputIt first, InputIt last) {
+    double outVolume = 0.0;
+    bool countSelfLoopsTwice = !G.isDirected();
+    count nodesInGraph = 0;
+    InputIt actIt = first;
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:154]: Probably it is cleaner to simply declare `actIt` in the for loops below. E.g., `for (auto actIt = first; actIt != last; ++actIt)`

File: include/networkit/graph/GraphTools.hpp [hunk 20]
```diff
@@ -129,13 +129,86 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+
+/**
+ * Returns the inVolume (sum of the in-degree of all nodes) of a range
+ * of nodes.
+ *
+ * @param G The input graph.
+ * @param first Iterator for the starting node in the set.
+ * @param last Iterator for the last node in the set.
+ * @return inVolume of the subgraph.
+ */
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:183]: Same comments as in `volume`, both code and documentation.

File: networkit/cpp/graph/GraphTools.cpp [hunk 21]
```diff
@@ -201,14 +201,17 @@ double density(const Graph &G) noexcept {
 }
 
 double volume(const Graph &G) {
-    if (G.isDirected())
-        throw std::runtime_error("Volume computation is only supported for undirected graphs.");
+    auto volume = 0.0;
```
Comment (angriman) [inline networkit/cpp/graph/GraphTools.cpp:204]: I don't think there is a need for `auto` here, simply use `double`.

File: networkit/cpp/graph/test/GraphToolsGTest.cpp [hunk 22]
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
...
+
+    auto doTest = [&](const Graph &G) {
+        // Volume for graph G
+        const auto volume = GraphTools::volume(G);
+
+        // Volume for either directed/undirected
+        double mod = directed() ? 1.0 : 2.0;
+        double baseG = weighted() ? static_cast<double>(G.totalEdgeWeight())
+                                  : static_cast<double>(G.numberOfEdges());
+        EXPECT_NEAR(volume, mod * baseG, 1e-7);
```
Comment (angriman) [inline networkit/cpp/graph/test/GraphToolsGTest.cpp:116]: What about `EXPECT_DOUBLE_EQ`?

File: networkit/graphtools.pyx [hunk 23]
```diff
@@ -337,6 +339,46 @@ cdef class GraphTools:
...
+		nodes : unordered_set
+			A set of nodes from the graph. 
+
+		Returns
+		-------
+		double
+			The inVolume of the input graph.
+		"""
+		return subgraphInVolume(graph._this, nodes)
+
```
Comment (angriman) [inline networkit/graphtools.pyx:381]: Could you please add some simple Python tests for these APIs?

**Review by angriman — CHANGES_REQUESTED** [review]

**Comment by fabratu** [comment]
Thanks for the additional input, took care of everything.

File: networkit/cpp/graph/test/GraphToolsGTest.cpp [hunk 24]
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
...
+        Graph::NodeIterator lastOfSet = Graph::NodeIterator(&G, seed);
+        Graph::NodeIterator last = Graph::NodeIterator(&G, G.upperNodeIdBound());
+
+        const auto volumeSet = GraphTools::volume(G, first, lastOfSet);
+        const auto inVolumeSet= GraphTools::inVolume(G, first, lastOfSet);
+        const auto volumeComplementSet = GraphTools::volume(G, lastOfSet, last);
+        const auto inVolumeComplementSet= GraphTools::inVolume(G, lastOfSet, last);
+
+        EXPECT_NEAR(volumeSet + volumeComplementSet, volume, 1e-7);
+        EXPECT_NEAR(inVolumeSet + inVolumeComplementSet, volume, 1e-7);
```
Comment (angriman) [inline networkit/cpp/graph/test/GraphToolsGTest.cpp:132]: I guess you can use `EXPECT_DOUBLE_EQ` here as well right?

File: include/networkit/graph/GraphTools.hpp [hunk 25]
```diff
@@ -129,13 +129,79 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+
+    // Compute volume of subgraph if the subset is really a subset
+    if (nodesInGraph != G.numberOfNodes()) {
+        for (auto actIt = first; actIt != last; ++actIt) {
+            outVolume += G.weightedDegree(*actIt, countSelfLoopsTwice);
+        }
+    // Otherwise compute volume of the complete graph
+    } else {
+        outVolume = volume(G);
+    }
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:167]: Does it really make sense to distinguish these two cases? I think that you can simply compute the volume as
```
for (auto actIt = first; actIt != last; ++actIt) {
    outVolume += G.weightedDegree(*actIt, countSelfLoopsTwice);
}
```
and we can assume that the input range is already a set, we should not check that.

File: include/networkit/graph/GraphTools.hpp [hunk 26]
```diff
@@ -129,13 +129,79 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+
+    // Compute volume of subgraph if the subset is really a subset
+    if (nodesInGraph != G.numberOfNodes()) {
+        for (auto actIt = first; actIt != last; ++actIt) {
+            inVolume += G.weightedDegreeIn(*actIt, countSelfLoopsTwice);
+        }
+    // Otherwise compute volume of the complete graph
+    } else {
+        inVolume = volume(G);
+    }
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:199]: Same thing here

File: networkit/cpp/graph/GraphTools.cpp [hunk 27]
```diff
@@ -201,14 +201,17 @@ double density(const Graph &G) noexcept {
...
+    double volume = 0.0;
 
-    if (!G.isWeighted()) {
-        return 2.0 * G.numberOfEdges();
+    if (G.isWeighted()) {
+        volume = G.totalEdgeWeight();
     } else {
-        return 2.0 * G.totalEdgeWeight();
+        volume = G.numberOfEdges();
+    }
```
Comment (angriman) [inline networkit/cpp/graph/GraphTools.cpp:210]: `G.totalEdgeWeight()` already handles the weighted/unweighted case, you can simply use `double volume = G.totalEdgeWeight()` and drop the `if`.

**Review by angriman — CHANGES_REQUESTED** [review]

File: networkit/cpp/graph/test/GraphToolsGTest.cpp (hunk 24 above)
Comment (fabratu) [reply networkit/cpp/graph/test/GraphToolsGTest.cpp:132]: No, there are differences in the range of approx. 1e-5

File: networkit/cpp/graph/test/GraphToolsGTest.cpp [hunk 28]
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
...
+
+        // Volume for either directed/undirected
+        double mod = directed() ? 1.0 : 2.0;
+        double baseG = weighted() ? static_cast<double>(G.totalEdgeWeight())
+                                  : static_cast<double>(G.numberOfEdges());
+        EXPECT_DOUBLE_EQ(volume, mod * baseG);
+ 
+        // Volume for subgraph G2
+        node seed = Aux::Random::integer(G.upperNodeIdBound());
+        auto sNodes = std::unordered_set<node>();
```
Comment (angriman) [inline networkit/cpp/graph/test/GraphToolsGTest.cpp:120]: This is never used

File: networkit/cpp/graph/test/GraphToolsGTest.cpp [hunk 29]
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
...
+
+        Graph::NodeIterator first = Graph::NodeIterator(&G, 0);
+        Graph::NodeIterator lastOfSet = Graph::NodeIterator(&G, seed);
+        Graph::NodeIterator last = Graph::NodeIterator(&G, G.upperNodeIdBound());
+
+        const auto volumeSet = GraphTools::volume(G, first, lastOfSet);
```
Comment (angriman) [inline networkit/cpp/graph/test/GraphToolsGTest.cpp:126]: This is not really needed, why don't you just use something like
```
const auto volumeSet = GraphTools::volume(G, G.nodeRange().begin(), G.nodeRange().begin() + seed);
```

File: include/networkit/graph/GraphTools.hpp [hunk 30]
```diff
@@ -129,13 +129,55 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+ * @return inVolume of the subgraph.
+ */
+template <class InputIt>
+double inVolume(const Graph &G, InputIt first, InputIt last) {
+    double inVolume = 0.0;
+
+    // Compute volume of subgraph
+    for (auto actIt = first; actIt != last; ++actIt) {
+        inVolume += G.weightedDegreeIn(*actIt, !G.isDirected());
+    }
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:174]: Even simpler:
```
for (; first != last; ++first)
    inVolume += G. weightedDegreeIn(*first, !G.isDirected());
```

File: include/networkit/graph/GraphTools.hpp [hunk 31]
```diff
@@ -129,13 +129,55 @@ std::pair<node, node> size(const Graph &G) noexcept;
...
+ * @return Volume of the graph.
+ */
+template <class InputIt>
+double volume(const Graph &G, InputIt first, InputIt last) {
+    double outVolume = 0.0;
+
+    // Compute volume of subgraph
+    for (auto actIt = first; actIt != last; ++actIt) {
+        outVolume += G.weightedDegree(*actIt, !G.isDirected());
+    }
```
Comment (angriman) [inline include/networkit/graph/GraphTools.hpp:154]: I wrote below a simpler version of this loop.

**Review by angriman — CHANGES_REQUESTED** [review]

File: networkit/cpp/graph/test/GraphToolsGTest.cpp [hunk 32]
```diff
@@ -101,6 +101,48 @@ TEST_P(GraphToolsGTest, testDensity) {
...
+    constexpr count n = 100;
+    constexpr double p = 0.1;
+
+    auto doTest = [&](const Graph &G) {
+        // Volume for graph G
+        const auto volume = GraphTools::volume(G);
+
+        // Volume for either directed/undirected
+        double mod = directed() ? 1.0 : 2.0;
+        double baseG = weighted() ? static_cast<double>(G.totalEdgeWeight())
```
Comment (angriman) [inline networkit/cpp/graph/test/GraphToolsGTest.cpp:114]: The type of `totalEdgeWeight` is `edgeweight` (double), you don't need to cast it to double.

**Review by angriman — DISMISSED** [review]

**Comment by angriman** [comment]
These changes need to be approved also by @michitux

**Review by michitux — DISMISSED** [review]
There are still some minor issues with the Python interface, see individual comments. Apart from that I approve this.

File: networkit/graphtools.pyx [hunk 33]
```diff
@@ -321,21 +323,47 @@ cdef class GraphTools:
...
+			A vector of nodes from the graph.
 
 		Returns:
 		--------
 		double
-			The volume of the input graph.
+			The volume of the subgraph.
 		"""
-		return volume(graph._this)
+		if nodes.size() == 0:
```
[the remark this answers is not in the document]
Comment (fabratu) [reply networkit/graphtools.pyx]: Thanks, yeah this would be cleaner. Adopted the changes (also the other comment).

**Review by angriman — APPROVED** [review]

File: networkit/graphtools.pyx [hunk 34]
```diff
@@ -321,21 +323,61 @@ cdef class GraphTools:
...
+			try:
+				cNodes = <vector[node]?>nodes
+				return volume[vector[node].iterator](graph._this, cNodes.begin(), cNodes.end())
+			except TypeError:
+				raise RuntimeError("Error, nodes must be a list of nodes.")
+		else:
+			return volume(graph._this)
+
+	@staticmethod
+	def inVolume(Graph graph, nodes = None):
```
Comment (michitux) [inline networkit/graphtools.pyx:356]: If for `inVolume`  the nodes parameter is not optional, why is it declared with a default value of `None`, which just raises an exception below?

Comment (fabratu) [reply networkit/graphtools.pyx]: True. Thanks for all the input and being more careful than me.

**Review by michitux — APPROVED** [review]

**Comment by michitux** [comment]
I see no further problems with this code. There is just one minor thing I would like to mention which can definitely wait for another round of improvements: Currently, the GIL is not released during the computation of the volume. This means that Python code in other threads cannot run in parallel. This would be possible though. What would be required is to first save the return value in a C++ variable, i.e., a variable declared, e.g.,  with `cdef double result`. Then the call to volume/inVolume could be wrapped in a `with nogil` block in which no Python objects are touched anymore and thus no GIL is required.

**Pull request merged**